source/python/idetect/nlp_models/category/category.py

```
@@ -0,0 +1,158 @@
+
+class Tokenizer(TransformerMixin):
+    def __init__(self, stop_words=None):
+        self.stop_words = stop_words
+        self.tokenizer = WordPunctTokenizer()
+        
+    def prepare_tokens(self, text, stop_words):
+        tokens = self.tokenizer.tokenize(text)
+        tokens = [t for t in tokens if len(t) > 2]
+        tokens = [t for t in tokens if t not in stop_words]
+        tokens = [t.lower() for t in tokens]
+        tokens = [t for t in tokens if not t.isdigit()]
+        return tokens
+    
+    def fit(self, X, *args):
+        return self
+    
+    def transform(self, X, *args):
+        X = X.map(lambda x: self.prepare_tokens(x, self.stop_words))
+        return X
+
+class TfidfTransformer(TransformerMixin):
+    def __init__(self, no_below=5, no_above=0.5, tfidf_model=None, dictionary=None):
+        self.dictionary = dictionary
+        self.tfidf_model = tfidf_model
+        self.no_below = no_below
+        self.no_above = no_above
+    
+    def make_dictionary(self, texts):
+        if not self.dictionary:
+            self.dictionary = gensim.corpora.Dictionary(texts)
+            if self.no_below or self.no_above:
+                self.dictionary.filter_extremes(no_below=self.no_below, no_above=self.no_above)
+        return self
+    
+    def tfidf_transform(texts, dictionary=None, tfidf_model=None):
+        if not dictionary:
+            dictionary = gensim.corpora.Dictionary(texts)
+        corpus = [dictionary.doc2bow(text) for text in texts]
+        if not tfidf_model:
+            tfidf_model = gensim.models.TfidfModel(corpus)
+        corpus_tfidf = tfidf_model[corpus]
+        return corpus_tfidf, dictionary, tfidf_model
+    
+    def make_corpus(self, texts):
+        corpus = [self.dictionary.doc2bow(text) for text in texts]
+        return corpus
+    
+    def make_tfidf_model(self, corpus):
+        if not self.tfidf_model:
+            self.tfidf_model = gensim.models.TfidfModel(corpus)
+        return self
```

I don't think you need to return anything here. The caller must already have `self`, since it called a method on it.

---

I think it's also a bit confusing that some of the `make_` methods return the thing they just made and others of them change the object. Should the mutators be called `set_tfidf_model`?

Also, it's surprising that a second call to `make_tfidf_model` with a different corpus silently does nothing. I would expect it to either replace the existing tfidf_model or to raise an exception to let the caller know that what they just tried didn't happen.

---

Think I've got most of these.